File: Data Structures & Algorithms/kth-smallest-integer-in-bst/submission-7.cpp
/**
 * Definition for a binary tree node.
 * struct TreeNode {
 *     int val;
 *     TreeNode *left;
 *     TreeNode *right;
 *     TreeNode() : val(0), left(nullptr), right(nullptr) {}
 *     TreeNode(int x) : val(x), left(nullptr), right(nullptr) {}
 *     TreeNode(int x, TreeNode *left, TreeNode *right) : val(x), left(left), right(right) {}
 * };
 */

class Solution {
   public:
    int kthSmallest(TreeNode* root, int k) {
        vector<int> result;

        // Brutally traverse through the BST and store every node's values in result
        auto dfs = [&](this auto&& self, TreeNode* node) -> void {
            if (node == nullptr) {
                return;
            }
            result.push_back(node->val);
            self(node->left);
            self(node->right);
        };

        dfs(root);

        // Sort result and return kth smallest element
        sort(result.begin(), result.end());
        return result[k - 1];
    }
};

File: Data Structures & Algorithms/valid-binary-search-tree/submission-7.py
# Definition for a binary tree node.
# class TreeNode:
#     def __init__(self, val=0, left=None, right=None):
#         self.val = val
#         self.left = left
#         self.right = right


class Solution:
    def isValidBST(self, root: Optional[TreeNode]) -> bool:
        # Return true if root is empty
        if not root:
            return True

        # Create a queue that carries current node and left/right bounds
        # and push root into the queue so we can start the while loop
        q = deque([(root, float("-inf"), float("inf"))])

        # Exit the while loop when queue is empty
        while q:
            # Get the front node and remove it from the queue
            (node, left, right) = q.popleft()

            # Continue is current node is empty because we don't need to process it
            if not node:
                continue

            # Return false if current node is not valid for any of left/right bounds
            if not (left < node.val < right):
                return False

            q.append((node.left, left, node.val))
            q.append((node.right, node.val, right))

        # Return true if we've successfully processed all nodes in the BST
        return True
